File: backend/app/routers/records.py
"""Medical records & digital intake.

File uploads (X-ray/MRI) should go directly to encrypted object storage via a
short-lived pre-signed URL; the API only stores a pointer (file_key) + checksum.
This keeps large PHI blobs out of the app servers. A presign stub is included.
"""
import uuid

from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..database import get_db
from ..models import MedicalRecord, Appointment, User, Role
from ..schemas import MedicalRecordCreate, MedicalRecordOut
from ..security import get_current_user

router = APIRouter(prefix="/api/records", tags=["records"])


@router.post("/upload-url")
def get_upload_url(filename: str, content_type: str = "application/octet-stream",
                   user: User = Depends(get_current_user)):
    """Return a (stubbed) pre-signed upload URL + the object key to store later.
    Replace with boto3 generate_presigned_url against an SSE-KMS bucket."""
    key = f"phi/{user.id}/{uuid.uuid4()}-{filename}"
    return {
        "upload_url": f"https://storage.orthocare.example/{key}?X-Amz-Signature=STUB",
        "file_key": key,
        "expires_in": 900,
    }


@router.post("", response_model=MedicalRecordOut, status_code=201)
def create_record(payload: MedicalRecordCreate,
                  user: User = Depends(get_current_user),
                  db: Session = Depends(get_db)):
    rec = MedicalRecord(patient_id=user.id, **payload.model_dump())
    db.add(rec)
    db.commit()
    db.refresh(rec)
    return rec


@router.get("/patient/{patient_id}", response_model=list[MedicalRecordOut])
def list_records(patient_id: str, user: User = Depends(get_current_user),
                 db: Session = Depends(get_db)):
    # Patients see only their own records; doctors/admins may view any.
    if user.role == Role.patient and user.id != patient_id:
        raise HTTPException(status_code=403, detail="Not allowed")
    return (db.query(MedicalRecord)
            .filter(MedicalRecord.patient_id == patient_id)
            .order_by(MedicalRecord.created_at.desc()).all())

File: server/routes/records.js
const { Router } = require("express");
const { v4: uuid } = require("uuid");
const db = require("../db");
const { authenticate } = require("../auth");

const router = Router();

router.post("/upload-url", authenticate, (req, res) => {
  const { filename, content_type } = req.query;
  const key = `phi/${req.user.id}/${uuid()}-${filename || "file"}`;
  res.json({
    upload_url: `https://storage.orthocare.example/${key}?X-Amz-Signature=STUB`,
    file_key: key,
    expires_in: 900,
  });
});

router.post("/", authenticate, (req, res) => {
  const { appointment_id, record_type, title, notes, medications, file_key, file_sha256 } = req.body;
  const id = uuid();
  db.prepare(
    `INSERT INTO medical_records (id, patient_id, appointment_id, record_type, title, notes, medications, file_key, file_sha256)
     VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?)`
  ).run(id, req.user.id, appointment_id || null, record_type || "intake", title || null, notes || null, medications || null, file_key || null, file_sha256 || null);

  const rec = db.prepare("SELECT * FROM medical_records WHERE id = ?").get(id);
  res.status(201).json(rec);
});

router.get("/patient/:patientId", authenticate, (req, res) => {
  if (req.user.role === "patient" && req.user.id !== req.params.patientId) {
    return res.status(403).json({ detail: "Not allowed" });
  }
  const records = db.prepare(
    "SELECT * FROM medical_records WHERE patient_id = ? ORDER BY created_at DESC"
  ).all(req.params.patientId);
  res.json(records);
});

module.exports = router;
